tests/python_package_test/test_basic.py

```diff
@@ -10,11 +10,30 @@
 from sklearn.datasets import load_breast_cancer, dump_svmlight_file, load_svmlight_file
 from sklearn.model_selection import train_test_split
 
+try:
+    from functools import lru_cache
+except ImportError:
+    warnings.warn("Could not import functools.lru_cache", RuntimeWarning)
+
+    def lru_cache(user_function, maxsize=None):
+        @wraps(user_function)
+        def wrapper(*args, **kwargs):
+            arg_key = tuple(args, [item for item in kwargs.items()])
+            if arg_key not in cache:
+                cache[arg_key] = user_function(*args)
+            return cache[arg_key]
+        return wrapper
```

Comment:
@jameslamb I think you can move this code to `compat.py` and reuse it in other files from `python_package_test`.

Reply:
I think we should avoid adding code to the library that is only used for tests.

* it can make `import lightgbm` slower for users
* it makes the library larger which can be a problem for cases like [AWS Lambda](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html), where there is a strict 250MB limit for the total size of all dependencies

I have an alternative suggestion...I just pushed https://github.com/microsoft/LightGBM/pull/3486/commits/25a9bc7736010eb4278583c576dc761599c0040d, which adds a `utils.py` that the tests can import from.

Reply:
Yeah, for sure, it will be better!